Sources/PackageLoading/TargetSourcesBuilder.swift

```diff
@@ -0,0 +1,311 @@
+        for declaredResource in target.resources {
+            let resourcePath = self.targetPath.appending(RelativePath(declaredResource.path))
+            if path.contains(resourcePath) {
+                if matchedRule != .none {
+                    diags.emit(.error("Duplicate rule \(declaredResource.rule) found for \(path)"))
+                }
+                matchedRule = declaredResource.rule.fileRule
+            }
+        }
+
+        // Match any sources explicitly declared in the manifest file.
+        if let declaredSources = target.sources {
+            for declaredSource in declaredSources {
+                let sourcePath = self.targetPath.appending(RelativePath(declaredSource))
+                if path.contains(sourcePath) {
+                    if matchedRule != .none {
+                        diags.emit(.error("Duplicate rule compile found for \(path)"))
+                    }
+                    matchedRule = .compile
+                }
+            }
+        }
+
+        // We haven't found a rule using that's explicitly declared in the manifest
+        // so try doing an automatic match.
+        if matchedRule == .none {
+            let effectiveRules: [FileRuleDescription] = {
+                // Don't automatically match compile rules if target's sources are
+                // explicitly declared in the package manifest.
+                if target.sources != nil {
+                    return self.rules.filter { $0.rule == .compile }
+                }
+                return self.rules
+            }()
+
+            if let needle = effectiveRules.first(where: { $0.match(path: path) }) {
+                matchedRule = needle.rule
+            }
+        }
+
+        return matchedRule
+    }
+
+    /// Compute the contents of the files in a target.
+    ///
+    /// This avoids recursing into certain directories like exclude or the
+    /// ones that should be copied as-is.
+    public func computeContents() -> [AbsolutePath] {
+        var contents: [AbsolutePath] = []
+        var queue: [AbsolutePath] = [targetPath]
+
+        while let curr = queue.popLast() {
+            // Ignore dot files.
+            if curr.basename.hasPrefix(".") { continue }
+
+            // Ignore xcodeproj and playground directories.
+            //
+            // FIXME: Ignore lproj directories until we add localization support.
+            if ["xcodeproj", "playgroun", "lproj"].contains(curr.extension) { continue }
+
+            // Ignore if this is an excluded path.
+            if self.excludedPaths.contains(curr) { continue }
+
+            // Append and continue if the path doesn't have an extension or is not a directory.
+            if curr.extension != nil || !fs.isDirectory(curr) {
+                contents.append(curr)
+                continue
+            }
+
+            // At this point, curr can only be a directory.
+
+            // The path is a symlinked directory. Warn and continue.
```

Comment:
why?

Reply:
I am not exactly sure why we do this (maybe to avoid symlink recursion?) but this is the existing behavior. We have a bug for this, I'll reference it here.